crates/core/src/operations/write.rs

```diff
@@ -1075,7 +1075,7 @@ impl std::future::IntoFuture for WriteBuilder {
                         actions.push(protocol.into())
                     }
 
-                    if schema != table_schema {
+                    if try_cast_batch(schema.fields(), table_schema.fields()).is_err() {
```

Comment:
Try cast batch is too lenient, you can already see it in the test failure.

Reply:
There are a couple issues that are worth digging into:

* [can_cast_types](https://docs.rs/datafusion/latest/datafusion/common/arrow/compute/kernels/cast/fn.can_cast_types.html#) seems like the wrong function to use in `try_cast_batch` considering it allows incompatible type downcasting, i.e. casting from [i64 to i8](https://docs.rs/arrow-cast/53.0.0/src/arrow_cast/cast/mod.rs.html#240).
* The failing rust test `test_issue_2105` has the `id` column type defined as `PrimitiveType::Integer`, which based on delta spec should map to 4 bytes int (i32). But the test [expects]( https://github.com/delta-io/delta-rs/blob/87010461cfe01563d91a4b9cd6fa468e2ad5f283/crates/core/tests/integration_datafusion.rs#L1126) the query result to have the type `ArrowDataType::Int64`. Is this expected? The original [MR](https://github.com/delta-io/delta-rs/pull/2106/files) that introduced this test has the expected type defined as `ArrowDataType::Int32`, which seems more reasonable, but it got changed to i64 in later schema casting MR.
* The failed python test `test_parse_stats_with_new_schema` is definitely a valid test failure that we should fix @PeterKeDer . Although it might be related to the same use of `can_cast_types` in `try_cast_batch` issue mentioned in my first point.